File: rogue-server/tracker/trackers/potion_tracker.py
from abc import ABC

from tracker.tracker_template import Tracker
from util.potion_effect_type import PotionEffectType

MONITORED_EFFECTS = [
    PotionEffectType.SPEED.value, PotionEffectType.SLOWNESS.value, PotionEffectType.JUMP_BOOST.value
]


class PotionTracker(Tracker, ABC):

    def __init__(self, data):
        super().__init__(data)

        self.pending_potions = []
        self.active_potions = []

    def handle(self, event):
        if event['type'] == 'out_entity_effect':
            packet = event['packet']

            if int(packet['entityId']) is not super().data.entity_id:
                return

            effect = packet['effectId']

            if effect not in MONITORED_EFFECTS:
                return

            potion = {
                'effectId': effect,
                'amplifier': packet['amplifier'],
                'duration': packet['duration'],
                'last_transaction': super().data.ping_tracker.last_transaction
            }
            self.pending_potions.append(potion)
        elif event['type'] == 'in_transaction':
            packet = event['packet']

            id = packet['transactionId']

            for potion in self.pending_potions:
                if potion['last_transaction'] == id:
                    self.pending_potions.remove(potion)

                    potion['end'] = super().data.ticks_existed + 20 * int((potion['duration']))
                    self.active_potions.append(potion)
        elif event['type'] == 'in_flying':
            for potion in self.active_potions:
                if super().data.ticks_existed >= potion['end']:
                    self.active_potions.remove(potion)
        elif event['type'] == 'out_remove_entity_effect':
            packet = event['packet']

            if int(packet['entityId']) is not super().data.entity_id:
                return

            for potion in self.active_potions:
                if potion['effectId'] == packet['effectId']:
                    self.active_potions.remove(potion)

    def get_potion_level(self, potion_type):
        for effect in self.active_potions:
            if effect['effectId'] == potion_type.value:
                return effect['amplifier']
        return 0

File: rogue-server/src/main/java/org/hostile/anticheat/tracker/impl/PotionTracker.java
package org.hostile.anticheat.tracker.impl;

import lombok.Getter;
import lombok.Setter;
import org.hostile.anticheat.data.PlayerData;
import org.hostile.anticheat.event.PacketEvent;
import org.hostile.anticheat.packet.inbound.WrappedPacketPlayInFlying;
import org.hostile.anticheat.packet.inbound.WrappedPacketPlayInTransaction;
import org.hostile.anticheat.packet.outbound.WrappedPacketPlayOutEntityEffect;
import org.hostile.anticheat.packet.outbound.WrappedPacketPlayOutRemoveEntityEffect;
import org.hostile.anticheat.tracker.Tracker;
import org.hostile.anticheat.util.PotionEffectType;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public class PotionTracker extends Tracker {

    private static final List<Integer> MONITORED_EFFECTS = Arrays.stream(PotionEffectType.values())
            .map(PotionEffectType::getPotionId)
            .collect(Collectors.toList());

    private final List<Potion> activePotions = new ArrayList<>();
    private final List<Potion> pendingPotions = new ArrayList<>();

    public PotionTracker(PlayerData data) {
        super(data);
    }

    @Override
    public void handle(PacketEvent event) {
        if (event.getPacket() instanceof WrappedPacketPlayOutEntityEffect) {
            WrappedPacketPlayOutEntityEffect packet = (WrappedPacketPlayOutEntityEffect) event.getPacket();

            if (!MONITORED_EFFECTS.contains((int) packet.getEffectId())) {
                return;
            }

            this.pendingPotions.add(new Potion(
                    PotionEffectType.of(packet.getEffectId()),
                    packet.getAmplifier(),
                    packet.getDuration(),
                    data.getPingTracker().getLastTransaction()
            ));
        } else if (event.getPacket() instanceof WrappedPacketPlayInTransaction) {
            WrappedPacketPlayInTransaction packet = (WrappedPacketPlayInTransaction) event.getPacket();

            short transactionId = packet.getTransactionId();

            this.pendingPotions.removeIf(potion -> {
                if (potion.getTransactionId() == transactionId) {
                    potion.setExpiresAt(data.getTicksExisted() + (20 * potion.getDuration()));
                    this.activePotions.add(potion);
                    return true;
                }

                return false;
            });
        } else if (event.getPacket() instanceof WrappedPacketPlayInFlying) {
            this.activePotions.removeIf(potion -> potion.getExpiresAt() > data.getTicksExisted());
        } else if (event.getPacket() instanceof WrappedPacketPlayOutRemoveEntityEffect) {
            WrappedPacketPlayOutRemoveEntityEffect packet = (WrappedPacketPlayOutRemoveEntityEffect) event.getPacket();

            if (data.getEntityId() != packet.getEntityId()) {
                return;
            }

            this.pendingPotions.add(new Potion(
                    PotionEffectType.of(packet.getEffectId()),
                    0,
                    0,
                    data.getPingTracker().getLastTransaction())
            );
        }
    }

    public int getPotionLevel(PotionEffectType potionEffectType) {
        return activePotions.stream()
                .filter(potion -> potion.getType().equals(potionEffectType))
                .map(Potion::getAmplifier)
                .findFirst()
                .orElse(0);
    }

    @Getter
    private static class Potion {

        private final PotionEffectType type;

        private final int amplifier;
        private final int duration;

        private final short transactionId;

        @Setter private int expiresAt;

        public Potion(PotionEffectType type, int amplifier, int duration, short transactionId) {
            this.type = type;
            this.amplifier = amplifier;
            this.duration = duration;
            this.transactionId = transactionId;
        }
    }
}
